Mathlib/CategoryTheory/DifferentialObject.lean

```diff
@@ -0,0 +1,368 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+
+! This file was ported from Lean 3 source module category_theory.differential_object
+! leanprover-community/mathlib commit 6876fa15e3158ff3e4a4e2af1fb6e1945c6e8803
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Int.Basic
+import Mathlib.CategoryTheory.Shift.Basic
+import Mathlib.CategoryTheory.ConcreteCategory.Basic
+
+/-!
+# Differential objects in a category.
+
+A differential object in a category with zero morphisms and a shift is
+an object `X` equipped with
+a morphism `d : X ⟶ X⟦1⟧`, such that `d^2 = 0`.
+
+We build the category of differential objects, and some basic constructions
+such as the forgetful functor, zero morphisms and zero objects, and the shift functor
+on differential objects.
+-/
+
+
+open CategoryTheory.Limits
+
+universe v u
+
+namespace CategoryTheory
+
+variable (C : Type u) [Category.{v} C]
+
+-- TODO: generalize to `HasShift C A` for an arbitrary `[AddMonoid A]` `[One A]`.
+variable [HasZeroMorphisms C] [HasShift C ℤ]
+
+/-- A differential object in a category with zero morphisms and a shift is
+an object `pt` equipped with
+a morphism `d : X ⟶ X⟦1⟧`, such that `d^2 = 0`. -/
+-- Porting note: Removed `@[nolint has_nonempty_instance]`
+structure DifferentialObject where
+  /-- The underlying object of a differential object. -/
+  pt : C
+  /-- The differential of a differential object. -/
+  d : pt ⟶ pt⟦(1 : ℤ)⟧
+  /-- The differential `d` satisfies that `d² = 0`. -/
+  d_squared' : d ≫ d⟦(1 : ℤ)⟧' = 0 := by aesop_cat -- Porting note: Originally `by obviously`
+#align category_theory.differential_object CategoryTheory.DifferentialObject
+
+-- Porting note : this resulted in a `INCORRECT DEF/LEMMA` liner error, so `Hom.comm` is
+-- made manually
+-- restate_axiom DifferentialObject.d_squared'
+
```

Comment:
```suggestion
```
We're removing uses of `restate_axiom`. You can just rename the field of the structure to remove the prime.